--- src/core.js	2014-03-10 19:24:11.000000000 -0400
+++ ../nvd3/src/core.js	2014-03-10 19:22:16.000000000 -0400
@@ -17,6 +17,33 @@
 nv.dispatch = d3.dispatch('render_start', 'render_end');
 
 // *************************************************************************
+// Function bind polyfill, from MDN
+// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Function/bind#Compatibility
+if (!Function.prototype.bind) {
+  Function.prototype.bind = function (oThis) {
+    if (typeof this !== "function") {
+      // closest thing possible to the ECMAScript 5 internal IsCallable function
+      throw new TypeError("Function.prototype.bind - what is trying to be bound is not callable");
+    }
+
+    var aArgs = Array.prototype.slice.call(arguments, 1),
+        fToBind = this,
+        fNOP = function () {},
+        fBound = function () {
+          return fToBind.apply(this instanceof fNOP && oThis
+                                 ? this
+                                 : oThis,
+                               aArgs.concat(Array.prototype.slice.call(arguments)));
+        };
+
+    fNOP.prototype = this.prototype;
+    fBound.prototype = new fNOP();
+
+    return fBound;
+  };
+}
+
+// *************************************************************************
 //  Development render timers - disabled if dev = false
 
 if (nv.dev) {
@@ -48,6 +75,11 @@
   return arguments[arguments.length - 1];
 };
 
+nv.deprecated = function(name) {
+  if (nv.dev && console && console.warn)
+    console.warn('`' + name + '` has been deprecated.');
+}
+
 
 nv.render = function render(step) {
   step = step || 1; // number of graphs to generate in each timeout loop
